### PR TITLE
Abstractification of LightGraphs

### PR DESCRIPTION
**Pull request by CarloLucibello**
With this PR I create the following structure for types in LightGraphs:

```julia
"""
Abstract Graph type
guarantees:
    vertices are integers in 1:nv(g)
    edges are pair of vertices

functions to implement:
    basic constructors
    edges(g) returns a set/interators on edges
    nv(g)
    ne(g)
    has_edge(g, e)
    fadj(g)
    copy(g, h)
    add_vertex!(g)
    rem_vertex!(g, v)
    add_edge!(g, e)
    rem_edge!(g, e)
"""
abstract SimpleGraph

"""
Abstract undirected graph type
"""
abstract Graph <: SimpleGraph

"""
Abstract directed graph type
"""
abstract DiGraph <: SimpleGraph

"""A type representing an undirected graph."""
type LightGraph <: Graph
    n::Int
    edges::Set{Edge}
    fadjlist::Vector{Vector{Int}} # [src]: (dst, dst, dst)
end

"""A type representing a directed graph."""
type LightDiGraph <: DiGraph
    n::Int
    edges::Set{Edge}
    fadjlist::Vector{Vector{Int}} # [src]: (dst, dst, dst)
    badjlist::Vector{Vector{Int}} # [dst]: (src, src, src)
end
```

This is a very structured abstraction, in particular vertices are required to be sequential integers and edges are pair of indeces. It requires for concrete types to define a small sets of core functions, and all the library builds on top of that. 

Notice that just two line of code had to be changed outside the core files, and those two lines where applying bad programming practices. No file in test/ was even touched

The core files have gone through some reorganization, which I think improves clarity and modularity.  

In summary, I think that this change highlights, focuses, and shapes the design decision that where taken with the creation of LightGraphs. Moreover it has absolutely no impact on ease of use, speed, and codestyle of LG, and it adds to its flexibility. This hopefully should address the rightful perplexities expressed in Seth's comment  #221. 

If this PR gets merged (by the way, docs should be improved), it will be easy to test new data structures (e.g. given edge iterators #221). Moreover if we inherit from Graph to obtain a base type for Network.jl  we immediately obtain all the goodness from LG for free JuliaGraphs/Networks.jl#1

Cheers, 
Carlo

### CODE REVIEW

**Comment by sbromberger** [comment]
Having toyed with LG abstractions in the past (see the sparsemx branch history for the ugly details), I'm not in favor of this change. There's also a history here (see https://github.com/JuliaGraphs/LightGraphs.jl/pull/72).

Ultimately, LG is not a graph factory. I didn't design it to be everything to everyone; it does one set of things very well, and we should be striving to improve that set without adding more implementations to the package.

**Comment by CarloLucibello** [comment]
LG is not a graph factory, I totally agree, and is founded on good design principles, in fact it should always come with one and only one graph implementation. 

This PR just adds one tiny layer of abstraction, which is very very structured, and basically just allows user to define types which can store some additional data besides the adjacency list.

What I see in #72 is way way ahead along the way of abstraction, it had impact on the way of thinking and of coding. On the contrary no single line of code we will think of and write down in the future will be different whether this PR gets merged or not.

 Also a sparsematrix type is not allowed in this abstraction, since it requires to store an adjacency list.

As I see it, this PR only adds functionality for the user and for LG, completely  for free and with no downside. The two immediate advantages I can see are the possibility to internally test the performance of new data structures, and quick and robust start for Networks.jl. Please I strongly ask to reconsider and to give this a try. In any case going back is as easy as changing a dozen line of codes without any API breacking changes

**Comment by sbromberger** [comment]
From an email exchange I had earlier this year with a couple of Julia core devs:

> First things first: I want the code to be honest, and I love that it’s attracted Julia experts to helping make it Julian. So - the union stuff is awesome, and that needs to go in post-haste. This was my first significant project in Julia, so I know it’s got some unorthodox things in it. I’m really excited to have expert eyes on the code (that includes you and @ngyeesian).

> As for the abstractions, I’m really torn. On the one hand, I want to say, “you can use LightGraphs for anything you want, but performance issues are on you” - and on the other hand I see lots of potential issues with that approach:

> 1) People will use it without understanding the implications and get frustrated because it’s so slow (as I did with Graphs.jl - admittedly, I was using it without understanding it, and that’s likely what caused the problems);
> 2) I really don’t want to have to troubleshoot someone else’s custom code;
> 3) Any modifications we make will have to take into account the fact that custom graph types might behave very differently, and either not implement that functionality or do something that we wouldn’t normally do;
> 4) People will add functionality for their particular custom type of graph, and that won’t really benefit the core of LightGraphs (which are simple graphs as defined in the README) since they won’t be defined for those graphs, and in any case trying to understand them so we can merge PRs will lead us back to #2;
> 5) Nobody seems to need it right now; and if they do,
> 6) Graphs.jl is designed specifically for that sort of work.

**Comment by jpfairbanks** [comment]
I'd just like to comment on the fact that there are at least two dimensions of abstraction for graphs. 

One is abstraction of the storage/access format. That is Adjacency List or  Incidence List or something more exotic. We definitely want to restrict only to Adjacency List representations because the mental gymnastics necessary to write fast code simultaneously for both Adjacency Lists and Incidence Lists are too tough for mere mortals.

The second dimension is abstraction in the sense of Julia Types. Right now LG uses a specific and concrete form of Adjacency List that is:
```julia
"""A type representing an undirected graph."""
type Graph
    vertices::UnitRange{Int}
    edges::Set{Edge}
    fadjlist::Vector{Vector{Int}} # [src]: (dst, dst, dst)
end
```
The vast majority of the code in the library depends only on the fact that this is an adjacency list for both correctness and performance. We also assume O(1) access to `has_edge(g,v)`. Features of an adjacency list are O(1) time random access to the neighborhoods, where the neighborhoods can be iterated over in O(deg) time. Almost none of the code relies on the specific types that make up the Graph type. It just relies on the performance and correctness of a few functions.

I agree with @sbromberger that it would be a bad idea to generalize Graph to the point where it captures both Adjacency List and Incidence List forms. But I agree with @CarloLucibello that we could generalize a little bit and use an `AbstractGraph` as an anchor for what it means to implement Adjacency List semantics.

**Comment by CarloLucibello** [comment]
thanks james, this clarifies to me the kind of abstraction I was seeking (one which enforced the Adjacency List semantics)

**Comment by CarloLucibello** [comment]
> Almost none of the code relies on the specific types that make up the Graph type. It just relies on the performance and correctness of a few functions.

None actually, as this PR shows

**Comment by sbromberger** [comment]
What's the specific use case for abstraction?

**Comment by jpfairbanks** [comment]
I think that @CarloLucibello would like other packages to inherit from AbstractGraph in order to let LightGraphs functions apply to their data structure. I am skeptical of this use but could be convinced of its utility. 

I like the idea of allowing people to support implicit graphs like 
```julia
type Grid
    dimension::Int
    sizes::Vector{Int}
end
```

Such a type would be able to be constant size storage but can provide the adjacency list semantics by generating neighborhood iterators on demand. For example 2d Grid has `fadj(grid,v) = [v-1, v+2 (sizes[1]-1)*v, (sizes[1]+1)*v]` or something like that.

Of course that would be outside the scope of LightGraphs.
I have implemented an implicit matrix for my research which uses highly structured graphs.

**Comment by sbromberger** [comment]
> I think that @CarloLucibello would like other packages to inherit from AbstractGraph in order to let LightGraphs functions apply to their data structure. 

This, to me, is a nonstarter since it then opens up a can of worms with respect to support issues. Right now LG is self-contained; that is, if someone has a problem with it, we can diagnose and fix very easily. If we allow derivatives in other packages via abstraction, then we get into the same set of problems that befell Graphs.jl - having to troubleshoot code that is not our own, with guarantees that may have been violated - and it will cause folks to think less of LG.

**Comment by CarloLucibello** [comment]
Here are the use case I can think of

1) Obviously the main reason one wants an abstract type is to be able to define concrete types inheriting the same behaviour. Here the situation is much different from that of Graphs.jl. We would have a much tighter 
abstraction, enforcing adjacency list semantics, and come with just one fully-fledged graph type. 
Most people will just use the library as it come, and only if someone needs more he would implement its own type, which, since adjacency semantics is enforced, will most probably will work just well. 
I cannot really see and horde of people coming here crying with their fancy data structures. And if it does, we simply turn abstracton off, it is easy and painless.
What really would discorauge users (and me against them)  and make them think less of LG, would be to not being able to fit it to their needs, just because some unnecessary constraints are imposed on it. 
We could also explicitely caution users against constructing their own types, if need be.
At this moment there are no real alternatives in julia to lightgraphs, so I think it should  intercept all the use cases it can, but still avoiding the adjacency list  vs incidence list gimnastic which, as james said, is what make codes complicated. 

2) graph views / filters. Those are common and usefeul features. I didn't look into how they are commonly implemented, but I think you go through some abstractions and adaptors

3) testing new data structures: at the present time it is a little hard to experiment with new data structures. To measure performance against the standard implementation, in the test I did with the graph iterators, I had to switch back and forth from master, comparing benchmarks was a painful things. If this gets merged instead, creating your new structure and benchmarking it against the standard implementation would be much faster 

4) Networks.jl. It could be the same case of 1), a user defined type, inheriting from AbstractGraph, but this would be under the umbrella of JuliaGraphs organization. If it grows up nicely it will accomodate most user need that are not intercepted by LG

not really a use case but:
5) design clarity: I realized what is the minimal set of functions that makes LG fast only once I started to separate the abstraction from the concrete type. Then james' comment on adjacency list semantic totally shapened the thing up.

**Comment by sbromberger** [comment]
Comments:

1) Abstract types in Julia don't have any behavior to inherit. All they have are "promises" that are implemented in the concrete types - promises that can (and will) be broken. There's nothing that would prevent someone from creating a MyGraph <: AbstractGraph and implementing anything he wants. Then when higher-level LightGraphs functions break (because there's no consistency in the underlying data structure, for example), we're going to be stuck dealing with 1) a disappointed user, and 2) a mess of code.

We've had this discussion already, and the decision was to keep things concrete due to the lack of traits. Graphs has kludged some notion of traits together; it works well for them but I don't want to go down that path.

I haven't heard anything yet that is different from the last time we debated this (keep in mind that I was on the other side back then, and some very core committers persuaded me that it was the wrong side to be on, so we removed abstraction).

**Comment by CarloLucibello** [comment]
I totally fail to see where we lose simplicity, elegance, and consistency. Also, I think that any analogy between this PR and the kind of abstraction attempted before here and in Graph.jl are inappropriate. The most abstract part of this PR is this discussion itself, which was  around an *abstract* abstract graph and not on *this* abstract graph. You are the maintener of the repo, yours is the responsability of keeping the right direction, and  you have done a great job with it 'til now, so maybe this is the right choice, but in this moment I fell I strongly disagree with it. Moreover, without this PR, I don't know where to start off with Networks.jl, the only viable alternative reamains some metaprogramming trick @jpfairbanks was proposing.  Anyway, let's close this and move on

**Comment by sbromberger** [comment]
@CarloLucibello I understand you're frustrated. Here's my suggestion for moving forward: come up with a set of goals/requirements/features for Networks.jl (create a roadmap issue, perhaps?) and then, when we have some **concrete** idea of what it is we want to do there, we can discuss what modifications to LG would be necessary to achieve them.

Right now changing some core aspects of LG to satisfy some yet-to-be-determined requirement of a newly-minted repo doesn't feel like the right thing to do, especially when the change contradicts discussions we've already had and decisions we've already made about exactly the same proposal.
